File: TestFiles/TestApplications/MVCApp/Services/BookService.cs
using MVCApp.Models;

namespace MVCApp.Services
{
    /// <summary>
    /// Service interface for book operations
    /// </summary>
    public interface IBookService
    {
        Task<IEnumerable<Book>> GetAllBooksAsync();
        Task<Book?> GetBookByIdAsync(int id);
        Task<IEnumerable<Book>> SearchBooksAsync(string searchTerm);
        Task<IEnumerable<Book>> GetBooksByGenreAsync(string genre);
        Task<IEnumerable<Book>> GetFeaturedBooksAsync();
        Task<IEnumerable<Book>> GetNewReleasesAsync();
        Task<IEnumerable<string>> GetAllGenresAsync();
        Task<IEnumerable<Book>> GetRelatedBooksAsync(int bookId);
    }

    /// <summary>
    /// In-memory implementation of book service
    /// </summary>
    public class BookService : IBookService
    {
        private readonly List<Book> _books = new();

        public BookService()
        {
            SeedSampleData();
        }

        public async Task<IEnumerable<Book>> GetAllBooksAsync()
        {
            await Task.Delay(10); // Simulate async operation
            return _books.Where(b => b.IsAvailable).ToList();
        }

        public async Task<Book?> GetBookByIdAsync(int id)
        {
            await Task.Delay(10);
            return _books.FirstOrDefault(b => b.Id == id && b.IsAvailable);
        }

        public async Task<IEnumerable<Book>> SearchBooksAsync(string searchTerm)
        {
            await Task.Delay(10);

            if (string.IsNullOrWhiteSpace(searchTerm))
                return await GetAllBooksAsync();

            return _books.Where(b => b.IsAvailable &&
                (b.Title.Contains(searchTerm, StringComparison.OrdinalIgnoreCase) ||
                 b.Author.Contains(searchTerm, StringComparison.OrdinalIgnoreCase) ||
                 b.Description.Contains(searchTerm, StringComparison.OrdinalIgnoreCase)))
                .ToList();
        }

        public async Task<IEnumerable<Book>> GetBooksByGenreAsync(string genre)
        {
            await Task.Delay(10);

            if (string.IsNullOrWhiteSpace(genre))
                return await GetAllBooksAsync();

            return _books.Where(b => b.IsAvailable &&
                b.Genre.Equals(genre, StringComparison.OrdinalIgnoreCase))
                .ToList();
        }

        public async Task<IEnumerable<Book>> GetFeaturedBooksAsync()
        {
            await Task.Delay(10);
            return _books.Where(b => b.IsAvailable).Take(6).ToList();
        }

        public async Task<IEnumerable<Book>> GetNewReleasesAsync()
        {
            await Task.Delay(10);
            return _books.Where(b => b.IsAvailable && b.PublishedDate >= DateTime.Now.AddMonths(-6))
                         .OrderByDescending(b => b.PublishedDate)
                         .Take(4)
                         .ToList();
        }

        public async Task<IEnumerable<string>> GetAllGenresAsync()
        {
            await Task.Delay(10);
            return _books.Where(b => b.IsAvailable)
                         .Select(b => b.Genre)
                         .Distinct()
                         .OrderBy(g => g)
                         .ToList();
        }

        public async Task<IEnumerable<Book>> GetRelatedBooksAsync(int bookId)
        {
            await Task.Delay(10);

            var book = await GetBookByIdAsync(bookId);
            if (book == null)
                return Enumerable.Empty<Book>();

            return _books.Where(b => b.IsAvailable && 
                                b.Id != bookId && 
                                b.Genre == book.Genre)
                         .Take(4)
                         .ToList();
        }

        private void SeedSampleData()
        {
            _books.AddRange(new[]
            {
                new Book { Id = 1, Title = "Clean Code", Author = "Robert C. Martin", ISBN = "978-0132350884", 
                          PublishedDate = new DateTime(2008, 8, 1), Genre = "Programming", 
                          Description = "A guide to writing readable, maintainable code.", Price = 45.99m, PageCount = 464 },
                
                new Book { Id = 2, Title = "Design Patterns", Author = "Gang of Four", ISBN = "978-0201633610", 
                          PublishedDate = new DateTime(1994, 10, 31), Genre = "Programming", 
                          Description = "Elements of reusable object-oriented software.", Price = 59.99m, PageCount = 395 },
                
                new Book { Id = 3, Title = "The Pragmatic Programmer", Author = "Andrew Hunt", ISBN = "978-0135957059", 
                          PublishedDate = new DateTime(2019, 9, 13), Genre = "Programming", 
                          Description = "Your journey to mastery, 20th Anniversary Edition.", Price = 49.99m, PageCount = 352 },
                
                new Book { Id = 4, Title = "Dune", Author = "Frank Herbert", ISBN = "978-0441172719", 
                          PublishedDate = new DateTime(1965, 1, 1), Genre = "Science Fiction", 
                          Description = "A science fiction epic of political intrigue and environmental themes.", Price = 16.99m, PageCount = 688 },
                
                new Book { Id = 5, Title = "1984", Author = "George Orwell", ISBN = "978-0451524935", 
                          PublishedDate = new DateTime(1949, 6, 8), Genre = "Dystopian Fiction", 
                          Description = "A dystopian social science fiction novel and cautionary tale.", Price = 13.99m, PageCount = 328 },
                
                new Book { Id = 6, Title = "To Kill a Mockingbird", Author = "Harper Lee", ISBN = "978-0061120084", 
                          PublishedDate = new DateTime(1960, 7, 11), Genre = "Fiction", 
                          Description = "A gripping tale of racial injustice and childhood innocence.", Price = 14.99m, PageCount = 281 },
                
                new Book { Id = 7, Title = "The Art of Computer Programming", Author = "Donald Knuth", ISBN = "978-0201896848", 
                          PublishedDate = new DateTime(1997, 7, 1), Genre = "Programming", 
                          Description = "Fundamental algorithms and data structures.", Price = 199.99m, PageCount = 650 },
                
                new Book { Id = 8, Title = "Introduction to Algorithms", Author = "Thomas H. Cormen", ISBN = "978-0262033848", 
                          PublishedDate = new DateTime(2009, 7, 31), Genre = "Programming", 
                          Description = "Comprehensive textbook on algorithms.", Price = 89.99m, PageCount = 1312 }
            });
        }
    }
}

File: TestFiles/SingleFiles/VB/BookService.vb
Imports System
Imports System.Collections.Generic
Imports System.Threading.Tasks

Namespace BookLibrary.Services

    ''' <summary>
    ''' Service for managing book operations in VB.NET
    ''' </summary>
    Public Class BookService
        Private ReadOnly _books As List(Of Book)

        Public Sub New()
            _books = New List(Of Book)()
            InitializeDefaultBooks()
        End Sub

        Public ReadOnly Property BookCount As Integer
            Get
                Return _books.Count
            End Get
        End Property

        Public Function GetAllBooksAsync() As Task(Of List(Of Book))
            Return Task.FromResult(New List(Of Book)(_books))
        End Function

        Public Async Function GetBookByIdAsync(id As Integer) As Task(Of Book)
            Await Task.Delay(10) ' Simulate async operation
            Return _books.FirstOrDefault(Function(b) b.Id = id)
        End Function

        Public Function SearchBooksAsync(searchTerm As String) As Task(Of List(Of Book))
            If String.IsNullOrWhiteSpace(searchTerm) Then
                Return GetAllBooksAsync()
            End If

            Dim filteredBooks = _books.Where(Function(b) 
                b.Title.Contains(searchTerm, StringComparison.OrdinalIgnoreCase) OrElse
                b.Author.Contains(searchTerm, StringComparison.OrdinalIgnoreCase) OrElse
                b.Description.Contains(searchTerm, StringComparison.OrdinalIgnoreCase)
            ).ToList()

            Return Task.FromResult(filteredBooks)
        End Function

        Public Function GetBooksByGenreAsync(genre As String) As Task(Of List(Of Book))
            If String.IsNullOrWhiteSpace(genre) Then
                Return GetAllBooksAsync()
            End If

            Dim filteredBooks = _books.Where(Function(b) 
                b.Genre.Equals(genre, StringComparison.OrdinalIgnoreCase)
            ).ToList()

            Return Task.FromResult(filteredBooks)
        End Function

        Public Async Function AddBookAsync(book As Book) As Task(Of Boolean)
            Try
                If book Is Nothing Then Return False
                
                book.Id = GetNextId()
                _books.Add(book)
                
                Await LogBookAddedAsync(book)
                Return True
            Catch ex As Exception
                Console.WriteLine($"Error adding book: {ex.Message}")
                Return False
            End Try
        End Function

        Public Function UpdateBookAsync(book As Book) As Task(Of Boolean)
            Dim existingBook = _books.FirstOrDefault(Function(b) b.Id = book.Id)
            If existingBook IsNot Nothing Then
                existingBook.Title = book.Title
                existingBook.Author = book.Author
                existingBook.Genre = book.Genre
                existingBook.Description = book.Description
                existingBook.Price = book.Price
                existingBook.PublishedYear = book.PublishedYear
                existingBook.PageCount = book.PageCount
                existingBook.CoverImageUrl = book.CoverImageUrl
                Return Task.FromResult(True)
            End If
            Return Task.FromResult(False)
        End Function

        Public Function DeleteBookAsync(id As Integer) As Task(Of Boolean)
            Dim bookToRemove = _books.FirstOrDefault(Function(b) b.Id = id)
            If bookToRemove IsNot Nothing Then
                _books.Remove(bookToRemove)
                Return Task.FromResult(True)
            End If
            Return Task.FromResult(False)
        End Function

        Private Function GetNextId() As Integer
            Return If(_books.Count = 0, 1, _books.Max(Function(b) b.Id) + 1)
        End Function

        Private Async Function LogBookAddedAsync(book As Book) As Task
            Await Task.Delay(5) ' Simulate logging operation
            Console.WriteLine($"Book added: {book.Title} by {book.Author}")
        End Function

        Private Sub InitializeDefaultBooks()
            _books.AddRange({
                New Book With {
                    .Id = 1,
                    .Title = "The Great Gatsby",
                    .Author = "F. Scott Fitzgerald",
                    .Genre = "Classic Literature",
                    .Description = "A classic American novel set in the Jazz Age.",
                    .Price = 12.99D,
                    .PublishedYear = 1925,
                    .PageCount = 180,
                    .CoverImageUrl = "/images/gatsby.jpg"
                },
                New Book With {
                    .Id = 2,
                    .Title = "To Kill a Mockingbird",
                    .Author = "Harper Lee",
                    .Genre = "Fiction",
                    .Description = "A gripping tale of racial injustice and loss of innocence.",
                    .Price = 14.99D,
                    .PublishedYear = 1960,
                    .PageCount = 324,
                    .CoverImageUrl = "/images/mockingbird.jpg"
                },
                New Book With {
                    .Id = 3,
                    .Title = "1984",
                    .Author = "George Orwell",
                    .Genre = "Dystopian Fiction",
                    .Description = "A dystopian social science fiction novel.",
                    .Price = 13.99D,
                    .PublishedYear = 1949,
                    .PageCount = 328,
                    .CoverImageUrl = "/images/1984.jpg"
                }
            })
        End Sub
    End Class

End Namespace
